docs/architecture/adr-023-protobuf-naming.md

```diff
@@ -0,0 +1,238 @@
+and [Buf](https://buf.build/docs/style-guide) builds upon that. To the
+extent possible, we want to follow industry accepted guidelines and wisdom for
+the effective usage of protobuf, deviating from those only when there is clear
+rationale for our use case.
+
+### Adoption of `Any`
+
+The adoption of `google.protobuf.Any` as the recommended approach for encoding
+interface types (as opposed to `oneof`) makes package naming a central part
+of the encoding as fully-qualified message names now appear in encoded
+transactions.
+
+### Current Directory Organization
+
+Thus far we have mostly followed [Buf's](https://buf.build) [DEFAULT](https://buf.build/docs/lint-checkers#default)
+recommendations, with the minor deviation of disabling [`PACKAGE_DIRECTORY_MATCH`](https://buf.build/docs/lint-checkers#file_layout)
+which although being convenient for developing code comes with the warning
+from Buf that:
+
+> you will have a very bad time with many Protobuf plugins across various languages if you do not do this
+
+### Adoption of gRPC Queries
+
+In [ADR 021](adr-021-protobuf-query-encoding.md), gRPC was adopted for Protobuf
+native queries. The full gRPC service path thus becomes a key part of ABCI query
+path. In the future, gRPC queries may be allowed from within persistent scripts
+by technologies such as CosmWasm and these query routes would be stored within
+script binaries.
+
+## Decision
+
+The goal of this ADR is to provide thoughtful naming conventions that
+
+* encourage a good user experience for when users interact directly with
+.proto files and fully-qualified protobuf names
+* balance conciseness against the possibility of either over-optimizing (making
+names too short and cryptic) or under-optimizing (just accepting bloated names
+with lots of redundant information)
+
+These guidelines are meant to act as a style guide for both the SDK and
+third-party modules.
+
+As a starting point, we should adopt all of the [DEFAULT](https://buf.build/docs/lint-checkers#default)
+checkers in [Buf's](https://buf.build) including [`PACKAGE_DIRECTORY_MATCH`](https://buf.build/docs/lint-checkers#file_layout),
+except:
+* [PACKAGE_VERSION_SUFFIX](https://buf.build/docs/lint-checkers#package_version_suffix)
+* [SERVICE_SUFFIX](https://buf.build/docs/lint-checkers#service_suffix)
+
+Further guidelines to be described below.
+
+### Principles
+
+#### Concise and Descriptive Names
+
+Names should be descriptive enough to convey their meaning and distinguish
+them from other names.
+
+Given that we are using fully-qualifed names within
+`google.protobuf.Any` as well as within gRPC query routes, we should aim to
+keep names concise, without going overboard. The general rule of thumb should
+be if a shorter name would convey more or else the same thing, pick the shorter
+name.
+
+For instance, `cosmos.bank.Send` (16 bytes) conveys roughly the same information
```

Comment:
I totally under the rationale and explanation, and in theory, I absolutely agree. But in this exact example, I'd much rather see `cosmos.bank.MsgSend` -- I think this would be a fair compromise. Essentially, the type (suffix) should provide as much context and meaning as possible while remaining concise as possible (i.e. `Send` vs `MsgSend`).

Reply:
So is your thinking that having the `Msg` prefix lets one quickly scan the types so see which ones are `Msg`s? I can see the benefit of that since this is a fairly common use case.

If we do stick with a prefix, I still would question the value of `Msg` vs some other prefix like `Op` or even `Tx`. `TxSend` might not be bad actually.

Reply:
`TxSend` would cause more confusion though, no? What's wrong with `Msg*` -- it is a message after all.

Reply:
> `TxSend` would cause more confusion though, no?

Maybe you're right...

> What's wrong with `Msg*` -- it is a message after all.

But message is pretty generic... and everything in protobuf is a message too. It feels even more confusing now that we have `message`s everywhere in protobuf lol...

Anyway, I don't have a really strong opinion here actually, I was just hoping we could come up with something a bit more elegant.

Reply:
As per our discussion today, I've update this to retain `Msg`